File: sdk/typescript/examples/monitor-jobs.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Example: Monitor job progress in real-time
 */

import { HyperSDK, JobStatus } from '../src';

function formatBytes(bytes: number): string {
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  let value = bytes;
  let unitIndex = 0;

  while (value >= 1024 && unitIndex < units.length - 1) {
    value /= 1024;
    unitIndex++;
  }

  return `${value.toFixed(2)} ${units[unitIndex]}`;
}

async function monitorJob(client: HyperSDK, jobId: string): Promise<number> {
  console.log(`Monitoring job: ${jobId}\n`);

  while (true) {
    try {
      const job = await client.getJob(jobId);

      // Clear line and print status
      process.stdout.write('\r\x1b[K'); // Clear current line
      process.stdout.write(`Status: ${job.status}`);

      if (job.progress) {
        const progress = job.progress;
        process.stdout.write(` | Progress: ${progress.percent_complete.toFixed(1)}%`);
        process.stdout.write(` | Phase: ${progress.phase}`);

        if (progress.current_file) {
          process.stdout.write(` | File: ${progress.current_file}`);
        }

        if (progress.bytes_downloaded && progress.total_bytes) {
          const downloaded = formatBytes(progress.bytes_downloaded);
          const total = formatBytes(progress.total_bytes);
          process.stdout.write(` | ${downloaded}/${total}`);
        }

        if (progress.estimated_remaining) {
          process.stdout.write(` | ETA: ${progress.estimated_remaining}`);
        }
      }

      // Check if job is complete
      if (
        job.status === JobStatus.COMPLETED ||
        job.status === JobStatus.FAILED ||
        job.status === JobStatus.CANCELLED
      ) {
        console.log(); // New line
        break;
      }

      await new Promise((resolve) => setTimeout(resolve, 2000)); // Poll every 2 seconds
    } catch (error: any) {
      console.error('\nError:', error.message);
      return 1;
    }
  }

  // Get final job state
  const job = await client.getJob(jobId);

  // Print final result
  console.log('='.repeat(80));
  if (job.status === JobStatus.COMPLETED) {
    console.log('✓ Job completed successfully!');
    if (job.result) {
      const result = job.result;
      console.log(`  VM Name: ${result.vm_name}`);
      console.log(`  Output Directory: ${result.output_dir}`);
      console.log(`  Total Size: ${formatBytes(result.total_size)}`);
      console.log(`  Duration: ${(result.duration / 1e9).toFixed(2)} seconds`);
      console.log(`  Export Method: ${result.export_method}`);
      console.log(`  Files:`);
      for (const file of result.files) {
        console.log(`    - ${file}`);
      }
    }
  } else if (job.status === JobStatus.FAILED) {
    console.log('✗ Job failed!');
    if (job.error) {
      console.log(`  Error: ${job.error}`);
    }
  } else if (job.status === JobStatus.CANCELLED) {
    console.log('⊘ Job was cancelled');
  }

  return 0;
}

async function monitorAllJobs(client: HyperSDK): Promise<number> {
  console.log('Monitoring all jobs...\n');

  while (true) {
    try {
      // Get all jobs
      const jobs = await client.listJobs(true);

      // Filter running jobs
      const runningJobs = jobs.filter((j) => j.status === JobStatus.RUNNING);

      if (runningJobs.length === 0) {
        console.log('No running jobs.');
        break;
      }

      // Clear screen
      process.stdout.write('\x1b[2J\x1b[H');

      // Print header
      console.log('='.repeat(80));
      console.log(`Running Jobs: ${runningJobs.length}`);
      console.log('='.repeat(80));
      console.log();

      // Print each job
      for (const job of runningJobs) {
        const jobId = job.definition.id || 'Unknown';
        const name = job.definition.name || job.definition.vm_path;
        console.log(`Job: ${jobId} - ${name}`);

        if (job.progress) {
          const progress = job.progress;
          const barLength = 40;
          const filled = Math.floor((barLength * progress.percent_complete) / 100);
          const bar = '█'.repeat(filled) + '░'.repeat(barLength - filled);
          console.log(`  [${bar}] ${progress.percent_complete.toFixed(1)}%`);
          console.log(`  Phase: ${progress.phase}`);

          if (progress.estimated_remaining) {
            console.log(`  ETA: ${progress.estimated_remaining}`);
          }
        }

        console.log();
      }

      await new Promise((resolve) => setTimeout(resolve, 3000)); // Refresh every 3 seconds
    } catch (error: any) {
      console.error('\nError:', error.message);
      return 1;
    }
  }

  return 0;
}

async function main() {
  const args = process.argv.slice(2);
  const jobId = args[0];
  const mode = jobId ? 'single' : 'all';

  // Initialize client
  const client = new HyperSDK('http://localhost:8080');

  try {
    // Login (if authentication is enabled)
    // await client.login('admin', 'password');

    if (mode === 'single') {
      return await monitorJob(client, jobId);
    } else {
      return await monitorAllJobs(client);
    }
  } catch (error: any) {
    console.error('Error:', error.message);
    return 1;
  }
}

main().then((exitCode) => process.exit(exitCode));

File: sdk/python/examples/monitor_jobs.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0

"""Example: Monitor job progress in real-time."""

import sys
import time
from transiva import HyperSDK, JobStatus


def format_bytes(bytes_val):
    """Format bytes to human-readable format."""
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if bytes_val < 1024.0:
            return f"{bytes_val:.2f} {unit}"
        bytes_val /= 1024.0
    return f"{bytes_val:.2f} PB"


def monitor_job(client, job_id):
    """Monitor a single job until completion."""
    print(f"Monitoring job: {job_id}\n")

    while True:
        try:
            job = client.get_job(job_id)

            # Clear line and print status
            print(f"\r\033[K", end="")  # Clear current line
            print(f"Status: {job.status.value}", end="")

            if job.progress:
                progress = job.progress
                print(f" | Progress: {progress.percent_complete:.1f}%", end="")
                print(f" | Phase: {progress.phase}", end="")

                if progress.current_file:
                    print(f" | File: {progress.current_file}", end="")

                if progress.bytes_downloaded and progress.total_bytes:
                    downloaded = format_bytes(progress.bytes_downloaded)
                    total = format_bytes(progress.total_bytes)
                    print(f" | {downloaded}/{total}", end="")

                if progress.estimated_remaining:
                    print(f" | ETA: {progress.estimated_remaining}", end="")

            # Check if job is complete
            if job.status in [JobStatus.COMPLETED, JobStatus.FAILED, JobStatus.CANCELLED]:
                print()  # New line
                break

            time.sleep(2)  # Poll every 2 seconds

        except KeyboardInterrupt:
            print("\n\nCancelling job...")
            client.cancel_job(job_id)
            print("Job cancelled.")
            return 1
        except Exception as e:
            print(f"\nError: {e}")
            return 1

    # Print final result
    print(f"\n{'='*80}")
    if job.status == JobStatus.COMPLETED:
        print("✓ Job completed successfully!")
        if job.result:
            result = job.result
            print(f"  VM Name: {result.vm_name}")
            print(f"  Output Directory: {result.output_dir}")
            print(f"  Total Size: {format_bytes(result.total_size)}")
            print(f"  Duration: {result.duration / 1e9:.2f} seconds")
            print(f"  Export Method: {result.export_method}")
            print(f"  Files:")
            for file in result.files:
                print(f"    - {file}")
    elif job.status == JobStatus.FAILED:
        print("✗ Job failed!")
        if job.error:
            print(f"  Error: {job.error}")
    elif job.status == JobStatus.CANCELLED:
        print("⊘ Job was cancelled")

    return 0


def monitor_all_jobs(client):
    """Monitor all running jobs."""
    print("Monitoring all jobs...\n")

    while True:
        try:
            # Get all jobs
            jobs = client.list_jobs(all=True)

            # Filter running jobs
            running_jobs = [j for j in jobs if j.status == JobStatus.RUNNING]

            if not running_jobs:
                print("No running jobs.")
                break

            # Clear screen
            print("\033[2J\033[H", end="")

            # Print header
            print(f"{'='*80}")
            print(f"Running Jobs: {len(running_jobs)}")
            print(f"{'='*80}\n")

            # Print each job
            for job in running_jobs:
                job_id = job.definition.id or "Unknown"
                name = job.definition.name or job.definition.vm_path
                print(f"Job: {job_id} - {name}")

                if job.progress:
                    progress = job.progress
                    bar_length = 40
                    filled = int(bar_length * progress.percent_complete / 100)
                    bar = '█' * filled + '░' * (bar_length - filled)
                    print(f"  [{bar}] {progress.percent_complete:.1f}%")
                    print(f"  Phase: {progress.phase}")

                    if progress.estimated_remaining:
                        print(f"  ETA: {progress.estimated_remaining}")

                print()

            time.sleep(3)  # Refresh every 3 seconds

        except KeyboardInterrupt:
            print("\nStopped monitoring.")
            return 0
        except Exception as e:
            print(f"\nError: {e}")
            return 1


def main():
    if len(sys.argv) > 1:
        job_id = sys.argv[1]
        mode = "single"
    else:
        job_id = None
        mode = "all"

    # Initialize client
    client = HyperSDK("http://localhost:8080")

    try:
        # Login (if authentication is enabled)
        # client.login("admin", "password")

        if mode == "single":
            return monitor_job(client, job_id)
        else:
            return monitor_all_jobs(client)

    finally:
        client.close()


if __name__ == "__main__":
    exit(main())
